var/spack/repos/builtin/packages/petsc/package.py

```diff
@@ -157,12 +157,15 @@ class Petsc(Package, CudaPackage, ROCmPackage):
     variant("kokkos", default=False, description="Activates support for kokkos and kokkos-kernels")
     variant("fortran", default=True, description="Activates fortran support")
 
-    # https://github.com/spack/spack/issues/37416
-    conflicts("^rocprim@5.3.0:5.3.2", when="+rocm")
-    # petsc 3.20 has workaround for breaking change in hipsparseSpSV_solve api,
-    # but it seems to misdetect hipsparse@5.6.1 as 5.6.0, so the workaround
-    # only makes things worse
-    conflicts("^hipsparse@5.6", when="+rocm @3.20.0")
+    with when("+rocm"):
+        # https://github.com/spack/spack/issues/37416
+        conflicts("^rocprim@5.3.0:5.3.2")
+        # hipsparse@5.6.0 broke hipsparseSpSV_solve() API, reverted in 5.6.1.
+        conflicts("^hipsparse@5.6.0", when="@:3.19")
+        # petsc@3.20 has workarounds for 5.6.0, but it incorrectly detects
+        # hipsolver 5.6.1 as 5.6.0, so now it works with the broken 5.6.0 but
+        # not with the fixed 5.6.1 release.
+        conflicts("^hipsparse@5.6.1", when="@3.20.0")
```

Comment:
```suggestion
        conflicts("^hipsparse@5.6.1)
```

with all the hip changes - its hard to get this correct for all petsc versions but this gives a bit more compatibility

3.20 should work with most hip versions (including 5.7.0) except 5.6.1
3.19 should work with 5.6.0, but not 5.6.1, 5.7.0
3.18 works with 5.2.0 - don't know if it works with higher versions or not. (it might work with 5.6.1 - but not 5.6.0)
3.18 and older? (for sure 5.1 - don't know higher)

Maybe its not worth retaining (accurate) hip compatibility for older versions...

Reply:
Hm, it shouldn't be that hard to add forward compat bounds, and it would in fact be helpful in case someone needs an older release for some reason.

Are you sure what you're saying is correct? What I understood:

- rocm 5.6.0 is breaking, 5.6.1 is reverts that. So for petsc < 3.20, the conflict is on 5.6.0 only;
- petsc 3.20.0 is the odd one out there, only working with 5.6.0 due to failed attempt to be compatible with both 5.6.0 and 5.6.1
- there's another breaking change in rocm 5.7, and compatibility for it only exists in petsc 3.20+

Reply:
Perhaps:

```suggestion
        conflicts("^hipsparse@5.6.1", when="@3.20:")
        conflicts("^hipsparse@5.6.1:", when="@3.19")
        conflicts("^hipsparse@5.6:", when="@:3.18")

```

Reply:
Ah, petsc 3.19 was released with a fix to deal with the breakage in 5.6.0, assuming that was there to stay?

Reply:
>  Ah, petsc 3.19 was released with a fix to deal with the breakage in 5.6.0, assuming that was there to stay?

yes - we moved on from 3.19 to 3.20 - so any fixes will now go to 3.20.x

To clarify:

> rocm 5.6.0 is breaking, 5.6.1 is reverts that. So for petsc < 3.20, the conflict is on 5.6.0 only;

well on the rocm side 5.6.0 is buggy - and 5.6.1 has the fix.

On the petsc side - 3.19 is fixed to work with 5.6.0 (and older versions). And 3.20 also works with it. And 3.20 is also fixed to work with 5.7.0. So the only conflict here is 5.6.1 (Due to our inability to detect it correctly). wrt 3.19/3.20

> petsc 3.20.0 is the odd one out there, only working with 5.6.0 due to failed attempt to be compatible with both 5.6.0 and 5.6.1

Not really - as describe above

> there's another breaking change in rocm 5.7, and compatibility for it only exists in petsc 3.20+

Well its more like - we fixed build issues in 3.20 wrt rocm-5.7.

Note: testing all versions of petsc with all rocm versions is not done (to see what works and what doesn't... For ex: we didn't test with 6.0.1 - And our test suite is mostly with  rocm-5.4.3 )..

also cc: @jczhang07

Reply:
I fixed the petsc hip version detection in https://gitlab.com/petsc/petsc/-/merge_requests/6909